channels/generic/websocket.py

```diff
@@ -1,20 +1,27 @@
 import json
 
+from asgiref.sync import async_to_sync
+
 from ..consumer import AsyncConsumer, SyncConsumer
-from ..exceptions import AcceptConnection, DenyConnection, StopConsumer
+from ..exceptions import AcceptConnection, DenyConnection, InvalidChannelLayerError, StopConsumer
 
 
 class WebsocketConsumer(SyncConsumer):
     """
     Base WebSocket consumer. Provides a general encapsulation for the
     WebSocket handling model that other applications can build on.
     """
+    groups = []
 
     def websocket_connect(self, message):
         """
         Called when a WebSocket connection is opened.
         """
-        # TODO: group joining
+        try:
+            for group in self.groups:
```

Comment:
what about accessing `self. groups` through a method that can be easily overridden (for cases where these groups might be dynamic based on the scope or maybe Mixins that have been added)

eg here call `self._get_groups()`

with the default implementation of '_get_groups`
```python
def _get_groups(self):
    return self.groups
```

starting the method name with `_` ensures that it cant be called through a channel layer message.

@andrewgodwin what do you think?

Reply:
That is what the `@property` decorator is for! It's still easy to override:

```
@property
def groups(self):
    ...
```